--- /d/Git/ta0829_3.2.2/triton-ascend/third_party/ascend/backend/backend_register.py	2026-08-29 20:29:54.727727000 +0800
+++ ./third_party/ascend/backend/backend_register.py	2026-08-29 16:45:54.983768200 +0800
@@ -309,11 +309,11 @@
 @backend_strategy_registry.register("torch_npu", "allocate_memory")
 def allocate_memory(size, stream):
     return f'''init_npu_utils();
-    if (!g_allocate_workspace_legacy) {{
-      fprintf(stderr, "Error: triton_allocate_workspace_legacy is unavailable\\n");
+    if (!g_allocate_workspace) {{
+      fprintf(stderr, "Error: triton_allocate_workspace is unavailable\\n");
       workspace_addr_ptr = nullptr;
     }} else {{
-      workspace_addr_ptr = g_allocate_workspace_legacy({size});
+      workspace_addr_ptr = g_allocate_workspace({size}, &workspace_handle);
     }}'''
 
 
